bigquery_etl/glam/templates/probe_counts_v1.udf.histogram.sql

```diff
@@ -0,0 +1,161 @@
+CREATE TEMP FUNCTION udf_exponential_buckets(min FLOAT64, max FLOAT64, nBuckets FLOAT64)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  let logMax = Math.log(max);
+  let current = min;
+  if (current === 0) {
+    current = 1;
+  } // If starting from 0, the second bucket should be 1 rather than 0
+  let retArray = [0, current];
+  for (let bucketIndex = 2; bucketIndex < Math.min(nBuckets, max, 10000); bucketIndex++) {
+    let logCurrent = Math.log(current);
+    let logRatio = (logMax - logCurrent) / (nBuckets - bucketIndex);
+    let logNext = logCurrent + logRatio;
+    let nextValue =  Math.round(Math.exp(logNext));
+    current = nextValue > current ? nextValue : current + 1;
+    retArray[bucketIndex] = current;
+  }
+  return retArray
+''';
+
+CREATE TEMP FUNCTION udf_linear_buckets(min FLOAT64, max FLOAT64, nBuckets FLOAT64)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  let result = [0];
+  for (let i = 1; i < Math.min(nBuckets, max, 10000); i++) {
+    let linearRange = (min * (nBuckets - 1 - i) + max * (i - 1)) / (nBuckets - 2);
+    result.push(Math.round(linearRange));
+  }
+  return result;
+''';
+
+CREATE TEMP FUNCTION udf_functional_buckets(
+  log_base INT64,
+  buckets_per_magnitude INT64,
+  range_max INT64
+)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  function sample_to_bucket_index(sample) {
+    // Get the index of the sample
+    // https://github.com/mozilla/glean/blob/master/glean-core/src/histogram/functional.rs
+    let exponent = Math.pow(log_base, 1.0/buckets_per_magnitude);
+    return Math.ceil(Math.log(sample + 1) / Math.log(exponent));
+  }
+
+  let buckets = new Set([0]);
+  for (let index = 0; index < sample_to_bucket_index(range_max); index++) {
+    // Avoid re-using the exponent due to floating point issues when carrying
+    // the `pow` operation e.g. `let exponent = ...; Math.pow(exponent, index)`.
+    let bucket = Math.floor(Math.pow(log_base, index/buckets_per_magnitude));
+    buckets.add(bucket);
+  }
+
+  return [...buckets]
```

Comment:
I tested this using [this query](
https://console.cloud.google.com/bigquery?sq=38084719550:3e0efcdce4a048aa937119866ef2204b
).

Effectively, calculate the distinct buckets for a distribution type. Then find the max reported value. Unnest the functional buckets created by this UDF. Finally, determine if the generated buckets overlap with all of the reported buckets. The result of this query are no rows.